src/components/CPU/BreakpointPanel.js

```diff
@@ -0,0 +1,167 @@
+import {ContextMenu, ContextMenuTrigger, MenuItem} from "react-contextmenu";
+import listeners from "../../utils/listeners";
+
+export default function BreakpointPanel(props) {
+	const [cpuBreakpoints, setCpuBreakpoints] = useState([]);
+	const [memoryBreakpoints, setMemoryBreakpoints] = useState([]);
+
+	const fetchBreakpoints = useCallback(() => {
+		props.ppsspp.send({
+			event: 'cpu.breakpoint.list',
+		}).then(({breakpoints}) => {
+			setCpuBreakpoints(breakpoints);
+		});
+		props.ppsspp.send({
+			event: 'memory.breakpoint.list',
+		}).then(({breakpoints}) => {
+			setMemoryBreakpoints(breakpoints);
+		});
+	}, [props.ppsspp]);
+
+	useEffect(() => {
+		fetchBreakpoints();
+		// TODO cpu.breakpoint.* not sent by PPSSPP?
+		const listeners_ = listeners.listen({
+			'cpu.stepping': () => {
+				fetchBreakpoints();
+			},
+			// TODO memory.breakpoint add/update/remove
+			'cpu.breakpoint.add': () => {
+				fetchBreakpoints();
+			},
+			'cpu.breakpoint.update': () => {
+				fetchBreakpoints();
+			},
+			'cpu.breakpoint.remove': () => {
+				fetchBreakpoints();
+			},
+		});
+		return () => listeners.forget(listeners_);
+	}, [fetchBreakpoints]);
+
+	return (
+		<React.Fragment>
+			<ContextMenuTrigger id="addBreakpoint" holdToDisplay={-1}>
+				<table className="BreakpointPanel__table">
+					<thead>
+						{/* TODO Left align header, first column sizing, row selection */}
+						<tr>
+							<th>&nbsp;</th>
+							<th>Type</th>
+							<th>Offset</th>
+							<th>Size/Label</th>
+							<th>Opcode</th>
+							<th>Condition</th>
+							<th>Hits</th>
+						</tr>
+					</thead>
+					<tbody>
+						{renderMemoryBreakpoints(memoryBreakpoints)}
+						{renderCpuBreakpoints(cpuBreakpoints)}
+					</tbody>
+				</table>
+			</ContextMenuTrigger>
+			{/*<BreakpointModal*/}
+			{/*	ppsspp={props.ppsspp}*/}
+			{/*	currentThread={props.currentThread}*/}
+			{/*	isOpen={false}*/}
+			{/*	// onClose={this.closeEditBreakpoint}*/}
+			{/*	// breakpoint={this.state.editingBreakpoint}*/}
+			{/*	// initialOverrides={this.state.creatingBreakpoint}*/}
+			{/*/>*/}
+			{renderAddBreakpointContextMenu()}
+			{renderEditBreakpointContextMenu()}
+		</React.Fragment>
+	);
+}
+
+function renderAddBreakpointContextMenu() {
+	return (
+		<ContextMenu id="addBreakpoint">
+			<MenuItem>
+				Add New
+			</MenuItem>
+		</ContextMenu>
+	);
+}
+
+function renderEditBreakpointContextMenu() {
+	// TODO probably can't do checked MenuItem so maybe render dynamic text depending on breakpoint state: Enable/Disable
+	return (
+		<ContextMenu id="editBreakpoint">
+			<MenuItem>
+				Enable
+			</MenuItem>
+			<MenuItem>
+				Edit
+			</MenuItem>
+			<MenuItem divider/>
+			<MenuItem>
+				Add New
+			</MenuItem>
+		</ContextMenu>
+	);
+}
+
+function renderMemoryBreakpoints(breakpoints) {
+	// TODO checkbox handler
+	return breakpoints.map((breakpoint, index) =>
+		<ContextMenuTrigger id="editBreakpoint" renderTag="tr" holdToDisplay={-1} key={index}>

```

Comment:
nit: It might be slightly better to use `breakpoint.address` as the key, since the goal is to keep the selection and prevent re-rendering.  I'm not sure that new items are always added at the end.

-[Unknown]

Reply:
I'm using a combinations of `address` and `size` now since you can have multiple breakpoints with same address.